cpp/simulations/2020_npis_sarscov2_wildtype_germany.cpp

```diff
@@ -541,34 +543,42 @@ mio::IOResult<void> run(RunMode mode, const fs::path& data_dir, const fs::path&
     //run parameter study
     auto parameter_study =
         mio::ParameterStudy<mio::osecir::Simulation<>>{params_graph, 0.0, num_days_sim, 0.5, size_t(num_runs)};
-    auto ensemble_results = std::vector<std::vector<mio::TimeSeries<double>>>{};
-    ensemble_results.reserve(size_t(num_runs));
-    auto ensemble_params = std::vector<std::vector<mio::osecir::Model>>{};
-    ensemble_params.reserve(size_t(num_runs));
-    auto save_result_result = mio::IOResult<void>(mio::success());
-    auto run_idx            = size_t(0);
-    parameter_study.run(
+    auto save_single_run_result = mio::IOResult<void>(mio::success());
+    auto ensemble = parameter_study.run(
         [](auto&& graph) {
             return draw_sample(graph);
         },
-        [&](auto results_graph) {
-            ensemble_results.push_back(mio::interpolate_simulation_result(results_graph));
+        [&](auto results_graph, auto&& run_idx) {
+            auto interpolated_result = mio::interpolate_simulation_result(results_graph);
 
-            ensemble_params.emplace_back();
-            ensemble_params.back().reserve(results_graph.nodes().size());
+            auto params = std::vector<mio::osecir::Model>{};
+            params.reserve(results_graph.nodes().size());
             std::transform(results_graph.nodes().begin(), results_graph.nodes().end(),
-                           std::back_inserter(ensemble_params.back()), [](auto&& node) {
+                           std::back_inserter(params), [](auto&& node) {
                                return node.property.get_simulation().get_model();
                            });
 
-            if (save_result_result && save_single_runs) {
-                save_result_result = save_result_with_params(ensemble_results.back(), ensemble_params.back(),
-                                                             county_ids, result_dir, run_idx);
+            if (save_single_run_result && save_single_runs) {
+                save_single_run_result =
+                    save_result_with_params(interpolated_result, params, county_ids, result_dir, run_idx);
             }
-            ++run_idx;
+            return std::make_pair(std::move(interpolated_result), std::move(params));
         });
-    BOOST_OUTCOME_TRY(save_result_result);
-    BOOST_OUTCOME_TRY(save_results(ensemble_results, ensemble_params, county_ids, result_dir, save_single_runs));
+
+    if (ensemble.size() > 0){
+        auto ensemble_results = std::vector<std::vector<mio::TimeSeries<double>>>{};
+        ensemble_results.reserve(ensemble.size());
+        auto ensemble_params = std::vector<std::vector<mio::osecir::Model>>{};
+        ensemble_params.reserve(ensemble.size());
+        for (auto&& run: ensemble)
+        {
+            ensemble_results.emplace_back(std::move(run.first));
+            ensemble_params.emplace_back(std::move(run.second));
+        }
+
+        BOOST_OUTCOME_TRY(save_single_run_result);
```

Comment:
maybe. BOOST_OUTCOME_TRY can only be used in functions that return IOResult themselves. The lambda does not return IOResult. Maybe it could. Then ParameterStudies::run would return IOResult as well. 

I think this way is OK. Very unlikely to fail writing and only minor data loss/wasted compute, error is reported correctly.